File: sdks/official/fraiseql-go/fraiseql/unset.go
package fraiseql

// Field represents a three-state field value for update mutations:
// Unset (not provided), Null (explicitly null), or Value (has a value).
// This mirrors the UNSET sentinel pattern used in the Python SDK.
type Field[T any] struct {
	value  T
	isSet  bool
	isNull bool
}

// Unset returns a Field in the "not provided" state.
// When serialized, unset fields are omitted entirely from the request.
func Unset[T any]() Field[T] { return Field[T]{} }

// Null returns a Field explicitly set to null.
// When serialized, null fields are included with a null value.
func Null[T any]() Field[T] { return Field[T]{isSet: true, isNull: true} }

// Value returns a Field with the given value.
func Value[T any](v T) Field[T] { return Field[T]{value: v, isSet: true} }

// IsUnset returns true if this field was not provided.
func (f Field[T]) IsUnset() bool { return !f.isSet }

// IsNull returns true if this field was explicitly set to null.
func (f Field[T]) IsNull() bool { return f.isSet && f.isNull }

// IsValue returns true if this field has a concrete value (not unset, not null).
func (f Field[T]) IsValue() bool { return f.isSet && !f.isNull }

// Get returns the value. Panics if the field is unset or null.
func (f Field[T]) Get() T {
	if !f.isSet || f.isNull {
		panic("fraiseql: cannot Get value from unset or null Field")
	}
	return f.value
}

// GetOr returns the value, or the provided default if the field is unset or null.
func (f Field[T]) GetOr(defaultVal T) T {
	if f.isSet && !f.isNull {
		return f.value
	}
	return defaultVal
}

File: sdks/official/fraiseql-python/src/fraiseql/unset.py
"""UNSET sentinel for three-state update mutation inputs.

In update mutations, fields can be in one of three states:

- **UNSET** — field was not provided; omit from generated SQL.
- **None** — explicitly set to NULL in the database.
- **<value>** — set to the given value.

This module provides the ``UNSET`` sentinel object, which is a singleton
that can be distinguished from ``None`` using ``is`` comparison.

Example::

    @fraiseql.input
    class UpdateUserInput:
        name: str | None | UNSET_TYPE = UNSET
        email: str | None | UNSET_TYPE = UNSET

    # In the generated SQL:
    #   name=UNSET  → field omitted entirely
    #   name=None   → SET name = NULL
    #   name="Alice" → SET name = 'Alice'
"""

from __future__ import annotations

from typing import Any


class _UnsetType:
    """Singleton sentinel type representing 'not provided'."""

    _instance: _UnsetType | None = None

    def __new__(cls) -> _UnsetType:
        if cls._instance is None:
            cls._instance = super().__new__(cls)
        return cls._instance

    def __repr__(self) -> str:
        return "UNSET"

    def __bool__(self) -> bool:
        return False

    def __eq__(self, other: Any) -> bool:
        return other is UNSET

    def __hash__(self) -> int:
        return hash("UNSET")


#: Sentinel value meaning "field not provided" in update mutations.
#: Compare with ``is``: ``if value is UNSET: ...``
UNSET: _UnsetType = _UnsetType()

# Re-export the type so callers can annotate: ``field: str | UNSET_TYPE``
UNSET_TYPE = _UnsetType
